### ISSUE TITLE
Shell Idea: Combine "App" and "Grain" pages

### ISSUE THREAD

**Issue by rchrd2** [sandstorm-io/sandstorm#1145]
I noticed a few threads about "App" and "Grain" pages in these issues, so I thought I'd throw this idea out there.

It seems like the "Apps" page and "Grains" page could be combined. As it is now, if you click an App icon, it shows you the grains in that app. 

I think ideally there would be one page called "Grains", that is a filesystem where you can organize all your content into folders. Then if you want to create a new "Grain" there would be a "New" button and that's where you would get the list of currently installed Apps. This is based off of how Google docs works. I think ultimately being able to organize grains into folders makes sense. 

I see Sandstorm as a personal computer in the cloud, so really developing a filesystem and advanced UX make sense to me. These are ideas I have been thinking about for a while, but Sandstorm is a project that has the most potential to realize these ideas. I found Sandstorm by Googling similar ideas and seeing what existed.

**Comment by ocdtrekkie** [comment]
I see Grains very much as the file system view (albeit minus folders, which we still badly need), and Apps more like a start menu/program launcher. Apps, as is common on desktop applications, also contains a file open dialog of applicable files. Both methods of navigating the virtual "desktop" that is your Sandstorm are valuable, depending on your usage. (Some people I've encountered don't realize that Word files exist outside of Word, whereas for someone like me, it's normal to look at things from the File Explorer perspective, and open what I want from there.)

Also, @rchrd2, if you check out some screenshots of Sandstorm's older UI, it actually was much more similar to that. There was a dropdown menu a long time ago that was the only way to create new files and open existing grains. I think where it's at now is better.

**Comment by ocdtrekkie** [comment]
As another example, you can actually right click in Windows Explorer, and go to New, and see some file types of things you can create there, but very few people use file explorers in that manner. Usually they open the app to create in, and then save it to a location later.

**Comment by rchrd2** [comment]
@ocdtrekkie Right click would be nice for power users, but a "New" button would be the default. ... I do seem to recall previous versions of Sandstorm had a "New" button.

Here are some screenshots of Google Drive for reference. I hid all my files, the right area would contain the filesystem:

![screen shot 2015-11-09 at 1 07 29 pm](https://cloud.githubusercontent.com/assets/182479/11046543/f1c831aa-86e2-11e5-8472-a2f82617a058.png)

- Click "Connected more apps"
![screen shot 2015-11-09 at 1 07 45 pm](https://cloud.githubusercontent.com/assets/182479/11046542/f1c7f3d4-86e2-11e5-82a1-f5a51ecb8cc3.png)

**Comment by ocdtrekkie** [comment]
I've used Drive before (I moved to Sandstorm!), so I know how their UI works, my question is why would it be better? The Apps tab is not much different from a full-screen New menu. (Though a minor cost of the app details page was adding an extra click to make a new file, there may be ways to fix that.)

**Comment by kentonv** [comment]
Another point to make here is that not everyone thinks hierarchical filesystems are the nicest way to organize data. Some people like spending the time to carefully organize their data into folders (I personally was insanely obsessive about this in the late 90's / early 00's), but some other people would rather be lazy, not organize at all, and instead count on having a reasonable set of tools available to find the thing they need when they need it. Sorting by app is one such tool that's easy for us to implement. Eventually I want to have stronger search features, but that's a bit more complicated.

Also some use cases call for a different kind of organization. Music libraries are often organized into playlists, which are very different from folders.

And of course folders vs. labels is basically a flamewar topic.

I think ideally the Sandstorm shell should not try to implement any of these, but should instead make the right APIs available so that apps can do the work, and everyone can then choose the UI they prefer.

**Comment by ocdtrekkie** [comment]
@kentonv I always thought of labels as just folders that were smart enough to let you put a single thing into multiple at a time.

**Comment by kentonv** [comment]
@ocdtrekkie That's what I thought too. But the gmail team spent a lot of years trying to convince people that they don't need "folders" because labels are strictly better, only to ultimately give up and implement folders as an option. So apparently there are people who strongly prefer folders, for some reason.

**Comment by ocdtrekkie** [comment]
Does labels do too much to break the illusion? I mean, everything in a file system is just a block of bits with an index entry saying "you can find it here". It's almost a somewhat arbitrary decision to state that block a bits constituting a file can only be found at one virtual location.

The biggest reason I'm concerned about having folders/labels/etc. handled by an app or API, is that I am worried that the shell will support that poorly. Or that it will be exceptionally hard to support that well.

**Comment by kentonv** [comment]
Perhaps a big reason people prefer folders is for hierarchy. Gmail eventually introduced a concept of "nested labels" but honestly I find that concept a little weird myself.

> The biggest reason I'm concerned about having folders/labels/etc. handled by an app or API, is that I am worried that the shell will support that poorly. Or that it will be exceptionally hard to support that well.

Having trouble parsing this. "support that poorly" -> what does "that" refer to? Are you saying you suspect an integration where an app takes over the "apps" and/or "grains" tabs would likely be janky?

**Comment by ocdtrekkie** [comment]
Yeah. I obviously don't want to have to open an app to access my other apps (and folders), and I don't want to have folders grains in addition to the existing grain list showing all the grains, so I'd guess it'd need to replace that tab... I'm totally cool with shell replacement being a supported feature for Sandstormm, and I could see a lot of potential options for what one might replace the shell with, but I kinda feel some sort of grain management should be supported officially.

Folders weren't a thing for Android homescreens for a long time, unless you got a third party launcher. I think we should all be able to agree that it's a better thing that the stock launcher now supports folders directly.

**Comment by ocdtrekkie** [comment]
If you have some basic support for say, labels, in the standard shell, one could easily use the labels field to manage "folders" in a custom version of the shell. And of course, people would also then be able to switch shells, and be reasonable comfortable that everything wouldn't be dumped back into one pile.
